File: test/scripts/10-blocking-io/correct.py
#!/usr/bin/env python3
"""
CORRECT: Async I/O - single thread, non-blocking.
Resume: "Use Asynchronous I/O"
"""
import asyncio
import os
import signal
import tempfile
import time
NUM_READS = 100
FILE_SIZE_KB = 100
DURATION_SEC = int(os.environ.get("DURATION_SEC", "300"))
stop = False

async def main_async():
    global stop
    print("CORRECT: Async I/O")
    with tempfile.NamedTemporaryFile(delete=False, suffix='.dat') as f:
        f.write(os.urandom(FILE_SIZE_KB * 1024))
        path = f.name
    try:
        start = time.perf_counter()
        laps = 0
        while not stop and (time.perf_counter() - start) < DURATION_SEC:
            def read_file(p):
                with open(p, 'rb') as f:
                    return len(f.read())
            loop = asyncio.get_event_loop()
            tasks = [loop.run_in_executor(None, read_file, path) for _ in range(NUM_READS)]
            await asyncio.gather(*tasks)
            laps += 1
        print(f"Stopped: {laps} laps")
    finally:
        os.unlink(path)

def main():
    global stop
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    signal.signal(signal.SIGTERM, lambda s, f: globals().__setitem__('stop', True))
    asyncio.run(main_async())

if __name__ == "__main__":
    main()

File: test/scripts/10-blocking-io/correct.js
const fs = require('fs').promises;
const path = require('path');
const os = require('os');
const NUM_READS = 100;
const FILE_KB = 100;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '300', 10);
const tmp = path.join(os.tmpdir(), 'perf-async-io-' + Date.now() + '.dat');
fs.writeFile(tmp, Buffer.alloc(FILE_KB * 1024)).then(() => {
  let laps = 0;
  const start = Date.now();
  console.log('CORRECT: Async reads');
  function run() {
    if (Date.now() - start >= DURATION_SEC * 1000) {
      fs.unlink(tmp).catch(() => {}).then(() => console.log('Stopped:', laps, 'laps'));
      return;
    }
    const tasks = Array(NUM_READS).fill(null).map(() => fs.readFile(tmp));
    Promise.all(tasks).then(() => {
      laps++;
      setImmediate(run);
    });
  }
  run();
});
process.on('SIGINT', () => process.exit(0));
